vsphere_plugin_common/__init__.py

```diff
@@ -596,139 +765,527 @@ def get_server_list(self):
+                    if cluster:
+                        ctx.logger.warn(
+                            'Host {host} is in cluster {cluster}, '
+                            'which is not an allowed cluster.'.format(
+                                host=host.name,
+                                cluster=cluster,
+                            )
+                        )
+                    else:
+                        ctx.logger.warn(
+                            'Host {host} is not in a cluster, '
+                            'and allowed clusters have been set.'.format(
+                                host=host.name,
+                            )
                         )
+                    continue
+
+            free_memory = self.get_host_free_memory(host)
+
+            if free_memory - vm_memory < 0:
+                ctx.logger.warn(
+                    'Host {host} does not have enough free memory.'.format(
+                        host=host.name,
                     )
-                    if dtstr_free_spc > selected_dtstr_free_spc:
-                        selected_datastore = datastore
-                        ctx.logger.debug(
-                            'Selected this datastore as a better candidate.'
+                )
+                continue
+
+            resource_pools = self.get_host_resource_pools(host)
+            resource_pools = [pool.name for pool in resource_pools]
+            if resource_pool not in resource_pools:
+                ctx.logger.warn(
+                    'Host {host} does not have resource pool {rp}.'.format(
+                        host=host.name,
+                        rp=resource_pool,
+                    )
+                )
+                continue
+
+            host_nets = set([
+                (
+                    network['name'],
+                    network['switch_distributed'],
+                )
+                for network in self.get_host_networks(host)
+            ])
+            vm_nets = set([
+                (
+                    network['name'],
+                    network['switch_distributed'],
+                )
+                for network in vm_networks
+            ])
+
+            nets_not_on_host = vm_nets.difference(host_nets)
+
+            if nets_not_on_host:
+                message = 'Host {host} does not have all required networks. '
+
+                missing_standard_nets = ', '.join([
+                    net[0] for net in nets_not_on_host
+                    if not net[1]
+                ])
+                missing_distributed_nets = ', '.join([
+                    net[0] for net in nets_not_on_host
+                    if net[1]
+                ])
+
+                if missing_standard_nets:
+                    message += 'Missing standard networks: {nets}. '
+
+                if missing_distributed_nets:
+                    message += 'Missing distributed networks: {dnets}. '
+
+                ctx.logger.warn(
```

Comment:
It's a str.format, and works fine- see example from interactive interpreter:
>>> test = 'Do something {hello}'
>>> test.format(hello='1', goodbye='2')
'Do something 1'